### ISSUE TITLE
White space editing

### ISSUE THREAD

**Issue by peteihis** [ArtOfIllusion/ArtOfIllusion#109]
Hi!

As anyone who has had a look at, what I have done so far can see, I have made quite a mess with the formatting of the code, primarily using tabs for indentation and occasionally panicking with the format and dropping to 2- or 4-space indents and the coming back.... I have been thinking, that before anyone does anything else, I should perform a "grand white wash" to the files, that I have created or edited but before that I'd like to make a suggestion about the formatting.

Now the guideline says, that the indentation style is 2-spaces, but I'm wondering, what the reasoning behind this instruction is or if the reasons are really valid nowadays?

I've got a feeling, that this practice pretty much comes from the era of VGA displays or before, where you could realistically fit about 20 lines on an editor screen and max 80 characters on a line. Nowadays displays are a lot larger. You can easily fit 60 x 200+ characters on the screen and with the latest toys even more.

While the 2-space indent saves space on small resolution displays, on large ones the presentation begins lose clarity of structure. At least that is very strongly my own perception of it. To me writing in 4-space indents makes the code much easier to browse and to keep track of the 'big picture'. Also (though this is really no argument here) Java examples on Oracle are written with 4-space indents and as I might guess, for very similar reasons.

Sure, there would be no point in starting to modify all of the code base into a new style but I'd suggest, that I'd clean/edit the new files (of my writing) and the ones, that I have heavily edited to 4-space indents. This would be rather easy to do even with files of currently mixed styles, while converting tabs to spaces.

While I'm at it, I might _(where I find it bothering me)_ do some other minor styling as well: In my experience for example this

~~~~java
    if(something)
      {
        doThis();
      }
    else
      {
        doThat();
      }
~~~~

is somewhat confusing to read especially, when the blocks get longer or when there are more similar indents inside the blocks. To me this would be more logical to follow:

~~~~java
    if(something)
    {
      doThis();
    }
    else
    {
      doThat();
    }
~~~~

Then especially in cases of if-without-else or a short(ish) for, I might go:

~~~~java
    for(something){
      doThis();
      doThat();
    }
~~~~

What works better seems to depend on, what is inside the block. At least, if there are more blocks, I'd probably want to drop the first brace on it's own line. The examples above are in 2-space style and currently you can find all of the three in different parts of the code, depending on when each piece was written.

So, what do you think? Would you find modernization to 4-space intents, with some other minor flavors reasonable in the mentioned cases or should I strictly stick to the old rules?

**Comment by peastman** [comment]
I've worked on projects that used all sorts of different formatting standards.  Some used two spaces, some four.  Some used end of line braces, others put them on the next line.  And so on.  My conclusion has been that there's no real "best" way of doing it.  You quickly get used to one style, and then it seems like the most readable one to you and all others look wrong.  The thing that's critical is to consistently use one style throughout the entire codebase.  Otherwise, it becomes much less readable.

For AoI we've standardized on two spaces for indent and braces on a separate line.  Also, always include a space after an if or for.  It should be `if (something)`, not `if(something)`.  I don't claim these choices are inherently better than other possibilities, but they're the ones we use throughout the code base.  Not sticking to them makes the code less readable.

**Comment by peteihis** [comment]
OK....

>  You quickly get used to one style, and then it seems like the most readable one to you and all others look wrong.

I wouldn't say that all others look wrong. It is pretty similar to driving a different car from the usual. You get adjusted in a few minutes _(and if you switch cars regularly, you stop noticing the difference)_, but that does not mean that some cars wouldn't be more ergonomic to drive than others.

I know, that I'm more sensitive to 'ergonomics' of written text than most other people _(have had more than a fair share of be-more-carefuls and learn-to-pay-attentions in my time)_. One could say, that my window of adjustment has it's limits in that sense. As strange as it may seem, I have found one of the limits somewhere between 2 or 4 spaces for indents.  With two I for example (and this really is just one example )regularly miss method ends and have to go back and have the machine highlight the other end. Good thing I can pick a red '}' quite easily :).  With four I have not noticed that difficulty. So, hence the suggestion -- if it works for me it probably works for most other people even better.

But I get the importance of standardization here and a kind of expected, that you not willing to widen the spec. I can live with that.

**Comment by peteihis** [comment]
One more detail in the instructions caught my eye. In the example on contribution guidelines, 
https://github.com/ArtOfIllusion/ArtOfIllusion/blob/master/docs/Contributing.md the example is:

~~~~java
public class Example
{
  public static String Braces = "Left, new line to start a block";
  public static int indentSpaces = 2;

  public void longParameterLists(boolean areStacked,
                                 String toPrevent,
				 int theLength,
				 String fromRunningInto,
				 int wrappingIssues)
  {
    if(youUseAConditionalWithAOneLiner)
       System.out.println("Just indent the statement to be executed!");

    for (Object obj: Collection group)
         obj.indent("One space past opening Parenthesis");
  }
}
~~~~

Shouldn't the `"One space past opening Parenthesis"` apply to the parameter list only? With the `if` and the `for above the opening parenthesis has its pair on the same line and the content should be with 2-space indent. -- That is the way the code has been written so far.

**Comment by peastman** [comment]
I didn't write that example. :)  Several errors in it jump out at me.

- There's a missing space between "if" and the opening parenthesis that immediately follows it.
- The indentation is inconsistent.
- It defines a field called `Braces`.  Field names should start with a lower case letter (except for constants, which can be all upper case).

**Comment by Sailsman63** [comment]
Thank You very much for the feedback!

RE: Code Guidelines:

- Capitalized `Braces` - My mistake, I'll fix that soon.
- Missing space after `if` - Ditto - I've always done this without the space, but agree that it should match with `for`, etc.
- Indents for `if` and `for`: These were intentional. What I intended to communicate is that:
  - In the case that the conditional/loop guards a _Single Statement_,
  - __And__ the statement is too long to fit on the same line as the condition,
  - __Then__ the statement may be indented one space past the parenthesis, just as the stacked argument list.

If we want the indent of that last case to match other standard indents (two space), then we almost need to insist on braces, just for visual feedback.

````
for (Object obj: Collection group)
{
  doTheThingWithACrazyLongName(obj, extra, arguments, here);
}
````
vs.
````
for (Object obj: Collection group)
  doTheThing(obj);
````
vs.
````
for (Object obj: Collection group)
     doTheThing(obj);
````
These are only for breaking up _single statements_ that are too long to do
````
for (Object obj: Collection group) doTheThing(obj);
````
~~~~~~~~~~~~~~

>that my window of adjustment has it's limits in that sense. As strange as it may seem, I have found one of the limits somewhere between 2 or 4 spaces for indents.

Good to know. Out of curiosity, what kind of editor do you use for code, and do you know if it uses a constant-width font, or a proportional font? (More print-ready font, punctuation takes up a little sliver, an 'o' takes up a bit more, a 'w' takes up even more) I've found that the later tend to make code, as such, harder to read because they de-emphasize manual indents.

You can see examples of both types on this page. The main text uses a proportional font, while the code blocks use a constant width font. (Old school computer font, all characters take up the same screen space.)

>I've got a feeling, that this practice pretty much comes from the era of VGA displays or before, where you could realistically fit about 20 lines on an editor screen and max 80 characters on a line. Nowadays displays are a lot larger. You can easily fit 60 x 200+ characters on the screen and with the latest toys even more.

You could. It would not be a good idea, though. Traditional line limits _were_ based on the fact that the VT-100 terminals of the time (Way before even VGA!) could only display 72 characters a line. They had inherited this from the Teletype terminals that they replaced. (To this day, *nix-y systems call their basic text interfaces tt0, tt1, etc, for 'TeleType device x' - this, even though the device is virtual, and even a simple RasberryPi can emulate dozens of them at a time)

However, this was less of a technical limit, and more due to the technology being limited by _ergonomic_ factors. As a line of text gets longer, it becomes more tiring to read. Lets take a copy of my local newspaper as an example. It's fairly typical of US print papers is this respect.

![dsc_0758](https://user-images.githubusercontent.com/4542515/53776943-5c500580-3ebd-11e9-8a05-481214875354.JPG)

It's about 35cm wide (14in). At it's print font size, it has room for about 200 characters, if there was only a single column. 

Too read such a line of text, your eyes would have to track the red line. Each time you have to move to a new line, you have to pause reading long enough to track back to the beginning, and hunt for the correct line. The same effect happens on a computer screen.

On the other hand, this paper is actually arranged in columns. Each is about 60 characters wide, and can be read by tracking the green line. Much easier, and less tiring.

Well designed modern web pages follow this rule as well. Both www.artofillusion.org and https://artofillusion.github.io/ (the replacement test site) limit text to about 80 characters.

On computers we can reclaim some of that horizontal screen space with context-relevant information that does not belong in the main file:

- Navigation sidebars
- images
- glossary entries relevant to the main text
- reference material that would normally be placed as footnotes/endnotes
- review a second document/source file along side the first
- (Programming specific) Display values of various variables in one pane, while we single-step through the program in another

TLDR; 72-80 Character line limit is still a relevant, modern goal.

~~~~~~~~~~~~~

On the other hand, we also have plenty of _vertical_ space to play with.

This is call K&R style braces:
````
for(something){
  doThis();
  doThat();
}
````
It was popularized by [*The C Programming Language*](https://en.wikipedia.org/wiki/The_C_Programming_Language), first edition, 1978. At the time, screens could only do ~40 lines at a time, so this was to save vertical space. (it also saved paper when the code listing was printed, which was not uncommon at the time)

Some of us call this style [Egyptian braces](https://blog.codinghorror.com/new-programming-jargon/). (scroll down to entry 3 to get the joke.) It's not my favorite. I like to see both the beginning and the end of the block delimited on the left at a quick scan, and our screens are now tall enough that they contain more data than is comfortable to fit in brain-space, even if we "waste" some lines on an opening brace.

**Comment by Sailsman63** [comment]
BTW, this:

````
if(something)
{
  doThis();
}
````
is the intended stye.

This:
if(something)
  {
    doThis();
  }
````
is not.

**Comment by peteihis** [comment]
The Egyptian made me think of Bangles.... 

Ah, it was the VT-100, I remember my math professor sitting at, though myself I got better acquainted with some early models of Wyse. :) – But yeah, I know about monospaced versus proportional fonts and why print media looks exactly the same everywhere up to using the same font.

The classic, virtually unchallenged, code editing font is, of course, Courier New, which (to my own surprise) I like to use in size 8 on a 1600x1200 screen  (1440 in height feels even better, but not usually available). I have noticed, that not only in graphics or 3D-work, the large display with high resolution also helps with coding. It just seems to provide the 'breathing space'. You can view the code with less scrolling and the relative scrolling distance looks smaller.  Lately Source Code Pro, that GitHub uses for code, makes serious competition to the Courier: It’s so good, that I rarely pay any attention to it, so it must be really good!

I suppose this would answer your editor question:

![editor](https://user-images.githubusercontent.com/22843957/53839731-f76ddb80-3fa0-11e9-80e9-cfb0f27034b5.png)

Of course, I am not suggesting to push 200 characters on one line _(Wikipedia is a disaster, when it comes to page layout)_, but above the occasional _(but only occasional an clearly structured)_ 100+-character lines don’t look so bad. On the contrary, some times I like to do something like

~~~~java

	BButton button1 = new BButton(”press me now”); button1.setEnabled(false);
	BButton button2 = new BButton(”press me too”); button2.setEnabled(false);

~~~~

typically , when there are several lines of similar pairs of statements. 

The idea is not to save space, but it actually reduces significantly the amount of visual information you have to process, when you just scroll/browse past the sequence. The thing is, that your brain automatically reads or at least tries to read every line that passes your eyes. And, when I say ”you”, I of course mean myself. Another benefit is, that you know exactly, where to find the second half of the pair. If they’d be on separate lines, you'd need probably three or four times as much processing to find the correct spot.

This is, I admit, a matter of learning but, in the case of single line `for` or `if` blocks, I’d definitely drop the parentheses. Again, not really for the saved space but they have no function. There was a time when I probably would have wanted the block markers to be there but my brain has at some point understood, that they are just visual trash, that have to be cleared from thinking. Less is definitely less, he seems to think... ;)

When it comes to typing/print fonts (OOT, I know), one of my all time favorites of proportional serif fonts would be Palatino Linotype, which I used as the base for this logo. Some graphic enhancement done to the font in InkScape. _(And sorry for the condition of the www-page the sticker refers to. :D)_

![musicatorsticker](https://user-images.githubusercontent.com/22843957/53840347-89c2af00-3fa2-11e9-8de3-6772d8a51297.png)

I think I have read a lot of books printed in PL. The character width is very nice, it looks pleasantly sophisticated and light to digest, but it has its issues:  It requires really good printing equipment to retain its characteristics and unfortunately the character spacing is a bit wobbly: Type "browse" in PL-font and the s looks like being a lot closer to e than w.

**Comment by Sailsman63** [comment]
Just pushed a fix to the typos, and (hopefully) clarified what the guide was trying to convey. Thoughts?

**Comment by peastman** [comment]
The text under the `for` and `if` lines should only be indented two spaces, not four.
